### PR TITLE
Add TypeScript support

### PR DESCRIPTION
**Pull request by jvhoven**
Dependencies:

- removed babel as an intermediary build step, rely on the TypeScript compiler for now.
- added and configured @typescript-eslint with prettier to enforce coding styleguide.
- added ts-jest for interoperability with Jest and TypeScript (configuration can be found in the package.json).
- updated jest to 24.5.0 resolve audit warnings.

API:

- introduced AbstractResource which can be used as a parent class to
remove redundancy of declaring `one`, `list`, etc. for type checking
purposes.

Documentation:

- removed flow types.
- altered existing examples to use AbstractResource.

---

### Thoughts
The `resource.ts` could use some refactoring and/or API changes. Currently I've retained all of the original functionality, which is of course necessary for backwards compatibility. This did however increase the amount of `@ts-ignore` necessary to keep the exact same functionality.

One possible introduction would be a [decorator](https://www.typescriptlang.org/docs/handbook/decorators.html) function along the lines of the following implementation:

```TypeScript

const Resource = <T extends { new (...args: any[]): {} }>(endpoint: string): Function => (entity: T): T => {
  const MergeClass = class {
    public async remove(): Promise<T> {
      const id = Reflect.get(this, 'id');

      if (id === undefined) {
        throw new Error('Cannot remove a Resource which has no id, this is a programmer error.');
      }

      await requestRemove(`${endpoint}/${id}`);
      Reflect.set(this, 'id', undefined);

      // `this` is not of type Pokemon, but we do want it to be Pokemon,
      // therefore we cast it by using the merge function.
      return merge(this);
    }

    public async save(): Promise<T> {
      const id = Reflect.get(this, 'id');
      let json = id ? await put(`${endpoint}/${id}`, this) : await post(`${endpoint}`, this);
      return merge(this, json);
    }

    public static async one(id: number, queryParams?: QueryParams): Promise<T> {
      const json = await get(`${endpoint}/${id}`, queryParams);
      return merge(this, json);
    }

    public static async list(queryParams?: QueryParams): Promise<T[]> {
      const list = await get(endpoint, queryParams);
      return list.map((properties: JSON) => {
        return makeInstance(this, properties);
      });
    }

    public static async page(queryParams?: QueryParams): Promise<Page<T>> {
      const page = await get(endpoint, queryParams);
      page.content = page.content.map((properties: JSON) => {
        return makeInstance(this, properties);
      });
      return page;
    }
  };
 
  // somehow merge but keep overwritten methods
};

// consumer.js
@Resource('api/pokemon')
class Pokemon extends AbstractResource<Pokemon> {
}
```

Some takeaways:

* AbstractResource would still be necessary for TypeScript intellisense.
* Merging the defined `const mergeClass = ...` with the consumer' defined class is difficult as there is an inversion of control (mergeClass preceeds Pokemon).
* Decorators as of yet are an experimental feature.
--- 

https://github.com/42BV/mad-spring-connect/pull/13 needs to be merged in order to pass the Jenkins test.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module.exports = {
```
Comment (Pouja) [inline .eslintrc.js]: You do not use the same eslint rules as https://github.com/42BV/redux-mad-authentication/pull/8/files#diff-e4403a877d80de653400d88d85e4801a

File: src/middleware.ts [hunk 2]
```diff
@@ -46,61 +44,52 @@ export class ErrorWithResponse extends Error {
...
+    return Promise.resolve({});
+  }
+  const contentType = response.headers.get('Content-Type');
+  if (contentType === null || contentType.includes('json') === false) {
+    throw new Error('mad-spring-connect: Content-Type is not json, will not parse.');
+  }
+  return response.json();
 }
+
+export default Middleware;
```
Comment (Pouja) [inline src/middleware.ts]: Why did you move the export to bottom line where all the other exports are inline?

File: .eslintrc.js (hunk 1 above)
Comment (jvhoven) [reply .eslintrc.js]: This project does not use React, therefore I've removed those rules.

File: src/middleware.ts (hunk 2 above)
Comment (jvhoven) [reply src/middleware.ts]: Out of habit, will change it to an inline export default.

**Comment by jvhoven** [comment]
@Pouja I've left a small example underneath the wall of text but I'll elaborate on it here:

```js
@Resource('api/pokemon')
class Pokemon extends AbstractResource<Pokemon> {
}
```

This would decorate the returning Pokemon class with the methods `save`, `remove` and the static methods `list`, `one`, `page`. Unfortunately it is still required to extend `AbstractResource` in order to have TypeScript completion (e.g. typing Pokemon(dot) will pop up `list`). 

The difficulty I've yet to solve in order to make this functional is:

> merging the defined const mergeClass = ... with the consumer' defined class is difficult as there is an inversion of control (mergeClass preceeds Pokemon).

Meaning there has to be a solution in place which enables user-defined methods to not get overwritten with the default implementation, while retaining the same merge process and returning the entity type, in this case Pokemon. However I'm sure this problem has been solved in some ORM library.

--- 
So do let me know if you want to put through with this as I would need to do a bit more research in order to get this functional. If you do have a better way of solving this (looking at you @MrHus), please do let us know!

File: src/middleware.ts (hunk 2 above)
Comment (jvhoven) [reply src/middleware.ts]: Actually, defining this in-line seems to be illegal syntax:

```js
export default type Middleware = (middleware: Promise<any>) => Promise<any>;
                 ^ ---- cannot find name 'type'
```

That's why it is defined at the bottom line.

**Comment by jvhoven** [comment]
Closed in favor of #17.
